<a href="https://colab.research.google.com/github/HannahParker/cse380-notebooks/blob/master/08_2_Ponder_and_Prove_The_RSA_Cryptosystem.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Ponder and Prove The RSA Cryptosystem
#### Due: Saturday, 27 February 2021, 11:59 pm.

## Use the Designated RSA Players


In the string in the cell below are encoded private **and** public keys, and the corresponding primes and exponents, all together dubbed "the designated RSA players". Your task is to figure out how to extract these numbers from that string, and use them for the following task.


In [ ]:
PEM = '''-----BEGIN RSA PRIVATE KEY-----
MIIEogIBAAKCAQEAzU6g/of54RvchTm5Qq0ec9Ve/BYikcYbsdvNk8/FFzhOicQ6
BseSOuvgbnuLxcJVW5gWgpOWqlu/tMOESquIoPbdnI1joIaVnDRfzzj5ZABuncnM
joU8FrD73q03LQILT6MgQBQKAY1BZnnfDikUe2cQXY/ZddVz2EexNejAd/qr1XlV
IPOPvZt5Irj23a366Hz/J0tbPmzdEeYZGJCDW4qPqG73Z4CKv+AN4K5vMNI2uRnJ
RXlvCGgNs2KM8WCDelzL0MNwEfIbxIjigA+zZ4AsHsfdGc4l9ukiytd463wscaPz
l2/A1g7rBd2jSDoN+FqgNQTDQKwRLGz76cLsTwIDAQABAoIBAAon+5rbyyaxpCUL
B/kea0U7puk9hxRDApe30eGgA5X0eR4jkONI+BjmFkIg5ncv990CFwr6lhdzVnZw
vZicB7Q+whO+gtEsFzaU+LBdlyi1RMOVegMK4EAXHT6UvwgA8+JKpYvF8gDFphiV
qs0ehx7bqFmYhh9oIcTvNvr9upe0vgN69jvTjrlnZRZ5LJazgtCPuShq0Y0cl3fi
u3idx3NzqMRLz5LJo6Os8QI7b2IhG4UhAuDAZimSxNGtWPIPGpeMqGrVDof+Ss7J
I4hS3V5gFIrIk+QuMa+r51kDVl+4EKqOCUDLEaG8kzttCRGampYZZs3D9bRnD51P
+Z7yJeECgYEA/XOnrI15kviSSyZRMxIU0B4f75nyiU+1Cch9ktF3ElIOMhz+FoGO
rQcqq5OV2qJqzO4+ixII1MdDNI0k4rdFVX0caXB6WXGBMxHE2zBxQzP6ixUAaWcG
JZA0QFUYMCtJ5NT/PxKW4LR9MaazgfZ6VEWw3+/DX5EXu7X/TNlkNskCgYEAz18O
pBeFKB+tOighZ0KCdr9sj03csRbR/tZDdmDmlWvJaqf66szlg5lkoRBCTa5spwyr
jGTFLesocxfbOiVBgSpR6rUiSd7B3MTGDnXOUISVCJH7W1FVW2C828Euq9h7hwy3
WxEPpIdqe05nNXvIjeWEQEHg8A31nZsZl8ai3lcCgYBNTLzS78MohA96RBF31gfr
AYUT+ovyPREmDrPd12zNdaFGv3jvPExbkVf+RGDr8aVJI1CH8dQnsS5aFMIvM14+
GI5VyixGo0uYW88CWt/wcyXyzVD21KkXQ8fr5wgdiNZcqGnAvtatad7VCdatyJK+
qRKs+d9IgmQOqA9ZTOQ3oQKBgB+PQfUrNzKyD3UIYn0KnDxiSa1NlkbFSFRWW5IK
kU0wSEkZI5DUeiGbGLuCc/TKlPKfdQQ62d9xKIjLmquwu9VikXD8/Cjt4+crc1EE
ENAkPWI+hViSekEb6eIv9mBk4/fbsZQEdrL9gPEfL4nuOsmNoqD85bTjCvxffHei
3WqlAoGAPkZAcTwNY6Yo5XdTFzh355ekB6SX+MFAQOOU+evYcPBhxS5Zk3SqkbaW
ENIxtkNfhpcb7H5ZGvegKyNrXj2cVg4B9q1KRakqt57Hl48dkntp/FsB8LHxXUm2
JSh2OHzk62JR/F7mlFk99QaSgHpImS4+To4mzYnqdyk1cOneVic=
-----END RSA PRIVATE KEY-----
'''

In [ ]:
#!pip install pycryptodome
from Crypto.PublicKey import RSA
key = RSA.importKey(PEM)
key

RsaKey(n=25917625865764614624507472751051103464428082986161080917325091381217018533586661301320669088349902902177733085775047172358486320715824906656398649217984618175709283947623838618438741621369178069406446397074531354687624880196917763646414868349698726051879996914421772243243204533332912677707592777579924307583300353717924471535363990028443240100883027021565162972138614363715385332043557983711232731557512311123057439497171145626527059340721442610742860541066026291672440695589372721662512134639537223476118733688390730619624279959344239164602032337463835460472300522928416214570267590796482170746195043253636122733647, e=65537, d=1282099318809357776716255346734023184333671743307356521262309020216146208796373894729414058996145463003497423807661365835882213890790001791050008709045365703734523987338396398995657420029584437813993610011377247232819321323808038050897615136330977460979217083427001321583143401392576756658498493750310734473412127994427246550627780142546002138546655543230546395

In [ ]:
exponent1 = key.d % (key.p - 1)
exponent2 = key.d % (key.q - 1)
print("Exponent1: ", exponent1)
print("Exponent2: ", exponent2)

Exponent1:  54281734801746268773390956884129058433932961006381595386384687760661548063534760199972236149996772086919138921396765523523135845827702342870788863283682531659902551286949195427860875056014181351792873302012992687842708845010954751896929245934634119977824444615975851427853805778752406613206531521561091127201
Exponent2:  22161904918224714227571032075881749386795312409771630363525238053118997722370080361584856221461208136566483883088425157810953246738075140732746844342501922443730736797754813676687653086499169951577612043029794169250409797584684823156941461334954559011460737190665568363878240562725251482827858964950659459749


**modulus:** 25917625865764614624507472751051103464428082986161080917325091381217018533586661301320669088349902902177733085775047172358486320715824906656398649217984618175709283947623838618438741621369178069406446397074531354687624880196917763646414868349698726051879996914421772243243204533332912677707592777579924307583300353717924471535363990028443240100883027021565162972138614363715385332043557983711232731557512311123057439497171145626527059340721442610742860541066026291672440695589372721662512134639537223476118733688390730619624279959344239164602032337463835460472300522928416214570267590796482170746195043253636122733647

**public exponent:**
65537

**private exponent:**
1282099318809357776716255346734023184333671743307356521262309020216146208796373894729414058996145463003497423807661365835882213890790001791050008709045365703734523987338396398995657420029584437813993610011377247232819321323808038050897615136330977460979217083427001321583143401392576756658498493750310734473412127994427246550627780142546002138546655543230546395502748472358091575440470145662211341028173724375798216436973128946136871189307605154819611405020453473033745994199655405599844873798447330929315593290467099660955817434589553928138613543786824912448451851412687420728126959767482757051304538388727023674849

**prime1 value:**
1779798906194739451972044*79753610471412080471556695548170777130366743840076039502662876748076963100473304963352590545433016597654993502523750394723485226239613519787056873595194902850137382499662419928837003477175462658073618318069595209725376131494946301812427316858866682753117975608975871345623801742505673

**prime2 value:**
145621091099427821970355196426414899695448605313736047637292312641594120085318624088348378001393944019065335296367166589879330552584041557870666727258326498014315349660563186678271377614788059065223777868863406904969330950903096977665577757319923494479055778350175391403999223156138440588539141065367091338839

**exponent1 value:** 
54281734801746268773390956884129058433932961006381595386384687760661548063534760199972236149996772086919138921396765523523135845827702342870788863283682531659902551286949195427860875056014181351792873302012992687842708845010954751896929245934634119977824444615975851427853805778752406613206531521561091127201

**exponent2 value:**
22161904918224714227571032075881749386795312409771630363525238053118997722370080361584856221461208136566483883088425157810953246738075140732746844342501922443730736797754813676687653086499169951577612043029794169250409797584684823156941461334954559011460737190665568363878240562725251482827858964950659459749

### Exploration of Exponent1 and Exponent2###

This is the website I found the formulas for the exponents on: 

https://crypto.stackexchange.com/questions/1729/why-does-the-pkcs1-rsa-private-key-structure-contain-more-than-just-exponent-and

## Test the Limits

Specifically, answer the question: What is the longest message that is a substring of the title page of the Book of Mormon (see below) that you can successfully encrypt and then decrypt again with the designated RSA players and the supplied code? Break any ties in length by preferring substrings that start and end on word boundaries, and express the most complete thought.

In [ ]:
title_page = 'THE BOOK OF MORMON AN ACCOUNT WRITTEN BY THE HAND OF MORMON UPON PLATES TAKEN FROM THE PLATES OF NEPHI WHEREFORE IT IS AN ABRIDGMENT OF THE RECORD OF THE PEOPLE OF NEPHI AND ALSO OF THE LAMANITES WRITTEN TO THE LAMANITES WHO ARE A REMNANT OF THE HOUSE OF ISRAEL AND ALSO TO JEW AND GENTILE WRITTEN BY WAY OF COMMANDMENT AND ALSO BY THE SPIRIT OF PROPHECY AND OF REVELATION WRITTEN AND SEALED UP AND HID UP UNTO THE LORD THAT THEY MIGHT NOT BE DESTROYED TO COME FORTH BY THE GIFT AND POWER OF GOD UNTO THE INTERPRETATION THEREOF SEALED BY THE HAND OF MORONI AND HID UP UNTO THE LORD TO COME FORTH IN DUE TIME BY WAY OF THE GENTILE THE INTERPRETATION THEREOF BY THE GIFT OF GOD AN ABRIDGMENT TAKEN FROM THE BOOK OF ETHER ALSO WHICH IS A RECORD OF THE PEOPLE OF JARED WHO WERE SCATTERED AT THE TIME THE LORD CONFOUNDED THE LANGUAGE OF THE PEOPLE WHEN THEY WERE BUILDING A TOWER TO GET TO HEAVEN WHICH IS TO SHOW UNTO THE REMNANT OF THE HOUSE OF ISRAEL WHAT GREAT THINGS THE LORD HATH DONE FOR THEIR FATHERS AND THAT THEY MAY KNOW THE COVENANTS OF THE LORD THAT THEY ARE NOT CAST OFF FOREVER AND ALSO TO THE CONVINCING OF THE JEW AND GENTILE THAT JESUS IS THE CHRIST THE ETERNAL GOD MANIFESTING HIMSELF UNTO ALL NATIONS AND NOW IF THERE ARE FAULTS THEY ARE THE MISTAKES OF MEN WHEREFORE CONDEMN NOT THE THINGS OF GOD THAT YE MAY BE FOUND SPOTLESS AT THE JUDGMENT SEAT OF CHRIST'

def abr(n, b):
    d = n // b
    m = n % b
    if (d == 0):
        return [m]
    else:
        l = abr(d, b)
        l.append(m)
        return l

def reconstruct_number_from_abr(list_of_coefficients, base):
    result = 0
    power = 1
    for coeff in list_of_coefficients:
        result += coeff * power
        power *= base
    return result

def from27to10(alphabetic_message):
    l = list(map(lambda n: int(ord(n) - ord('@')), alphabetic_message))
    l.reverse()
    return reconstruct_number_from_abr(l, 27)

def to27from10(decimal_number):
    l = list(map(lambda n: chr(n + ord('@')), abr(decimal_number, 27)))
    return ''.join(l)

def igcd(b, n):
    x0, x1, y0, y1 = 1, 0, 0, 1
    while n != 0:
        q, b, n = b // n, n, b % n
        x0, x1 = x1, x0 - q * x1
        y0, y1 = y1, y0 - q * y1
    return b, x0, y0

def TUMMI(e, t):
    hcf, x, y = igcd(e, t)
    if hcf != 1:
        raise ValueError(f"TUMMI does not exist for e = {e} and t = {t}.")
    return x % t

def RSAencrypt(m, e, n, encoder=lambda x: x):
    message_encoded = encoder(m)
    encrypted = pow(message_encoded, e, n)
    return encrypted

def RSAdecrypt(encrypted, d, n, decoder=lambda x: x):
    decrypted = pow(encrypted, d, n)
    message_decoded = decoder(decrypted)
    return message_decoded

RSAplayers = {'p': 177979890619473945197204479753610471412080471556695548170777130366743840076039502662876748076963100473304963352590545433016597654993502523750394723485226239613519787056873595194902850137382499662419928837003477175462658073618318069595209725376131494946301812427316858866682753117975608975871345623801742505673,
              'q': 145621091099427821970355196426414899695448605313736047637292312641594120085318624088348378001393944019065335296367166589879330552584041557870666727258326498014315349660563186678271377614788059065223777868863406904969330950903096977665577757319923494479055778350175391403999223156138440588539141065367091338839,
              }

def populateRSAplayers(p, q, e = 2 ** 16 + 1):
    global RSAplayers
    n = p * q
    t = (p - 1) * (q - 1)
    d = TUMMI(e, t)
    RSAplayers['p'] = p
    RSAplayers['q'] = q
    RSAplayers['n'] = n
    RSAplayers['t'] = t
    RSAplayers['e'] = e
    RSAplayers['d'] = d

def round_trip(message, n, e, d, encoder=lambda x: x, decoder=lambda x: x):
    return message == RSAdecrypt(RSAencrypt(message, e, n, encoder), d, n, decoder)

def testRSA(message_with_spaces):
    global RSAplayers
    message = '@'.join(message_with_spaces.split(' '))
    n = RSAplayers['n']
    e = RSAplayers['e']
    d = RSAplayers['d']
    encrypted = RSAencrypt(message, e, n, from27to10)
    decrypted = RSAdecrypt(encrypted, d, n, to27from10)
    decrypted_message_with_spaces = ' '.join(decrypted.split('@'))
    success = (message_with_spaces == decrypted_message_with_spaces)
    #print('"{}" was encoded and encrypted as\n\n{}\n\nthen decrypted and decoded as\n\n"{}"\n'.
     #     format(message_with_spaces, encrypted, decrypted_message_with_spaces))
    #print('{}successfully.'.format('' if success else 'un'))
    return success

In [ ]:
populateRSAplayers(key.p, key.q)

In [ ]:
for i in range(1,1000):
  if not testRSA(title_page[0:i]):
    print(i)
    break

431


In [ ]:
for i in range(1,len(title_page) - 431):
  if title_page[i-1] == ' ' and title_page[i+431] ==' ':
    if testRSA(title_page[i:i+431]):
      print(title_page[i:i+431])

BOOK OF MORMON AN ACCOUNT WRITTEN BY THE HAND OF MORMON UPON PLATES TAKEN FROM THE PLATES OF NEPHI WHEREFORE IT IS AN ABRIDGMENT OF THE RECORD OF THE PEOPLE OF NEPHI AND ALSO OF THE LAMANITES WRITTEN TO THE LAMANITES WHO ARE A REMNANT OF THE HOUSE OF ISRAEL AND ALSO TO JEW AND GENTILE WRITTEN BY WAY OF COMMANDMENT AND ALSO BY THE SPIRIT OF PROPHECY AND OF REVELATION WRITTEN AND SEALED UP AND HID UP UNTO THE LORD THAT THEY MIGHT
BY THE HAND OF MORMON UPON PLATES TAKEN FROM THE PLATES OF NEPHI WHEREFORE IT IS AN ABRIDGMENT OF THE RECORD OF THE PEOPLE OF NEPHI AND ALSO OF THE LAMANITES WRITTEN TO THE LAMANITES WHO ARE A REMNANT OF THE HOUSE OF ISRAEL AND ALSO TO JEW AND GENTILE WRITTEN BY WAY OF COMMANDMENT AND ALSO BY THE SPIRIT OF PROPHECY AND OF REVELATION WRITTEN AND SEALED UP AND HID UP UNTO THE LORD THAT THEY MIGHT NOT BE DESTROYED TO COME FORTH BY
ARE A REMNANT OF THE HOUSE OF ISRAEL AND ALSO TO JEW AND GENTILE WRITTEN BY WAY OF COMMANDMENT AND ALSO BY THE SPIRIT OF PROPHECY AND OF

**Answer:**

AT THE TIME THE LORD CONFOUNDED THE LANGUAGE OF THE PEOPLE WHEN THEY WERE BUILDING A TOWER TO GET TO HEAVEN WHICH IS TO SHOW UNTO THE REMNANT OF THE HOUSE OF ISRAEL WHAT GREAT THINGS THE LORD HATH DONE FOR THEIR FATHERS AND THAT THEY MAY KNOW THE COVENANTS OF THE LORD THAT THEY ARE NOT CAST OFF FOREVER AND ALSO TO THE CONVINCING OF THE JEW AND GENTILE THAT JESUS IS THE CHRIST THE ETERNAL GOD MANIFESTING HIMSELF UNTO ALL NATIONS


$\color{red}{\text{Correct!}}$

## What is True?
Assess yourself on how you did using the checkboxes below. Check a box by putting an 'X' in it only if it is warranted.


# TODO My Report on What I Did and What I Learned

## Fun

There were parts of this assignment that were fun and other parts that I really didn't like. I thought that extracting the RSA players was fun. I also thought that finding the string was fun. However, I did not enjoy how long this took me. It took awhile to run and rerun each code block to find the correct answer. 

## New

I learned how to extract from RSA keys. This was new to me. I learned about the different crytography packages in Python as well. Together, this helped me to expand my knowledge on RSA and what the RSA players do specifically. 

## Meaningful

I found this task meaningful because I learned a lot of new informantion. If I ever have to use RSA encryption again, I will be thankful for this knowledge I gained. I feel like I have a good understanding of what each of the RSA players do now. I also feel that this was meaningful because I learned more about writing code in python. 

## Other

Connections: I can connect this to the concepts I learned in my Intro to Cybersecurity class last semester. In that class I learned a lot about password encryption. This connects to RSA because it is a type of encryption. Learning about RSA helped me to understand how this actually works.

Collaboraters: Hunter Livesay

# TODO What is True?
Click on each warranted checkbox to toggle it to True (or back to False). 

NOTE: *This only works in Colab. If you run it in some other Jupyter notebook client/server environment you may have to change False to True (or vice versa) manually.*

This self-assessment is subject to revision by a grader.

In [ ]:
#@markdown ## What is True about what I did?
#@markdown ### I had fun.
cb00 = True #@param {type:'boolean'}
#@markdown ### I learned something new.
cb01 = True #@param {type:'boolean'}
#@markdown ### I achieved something meaningful, or something I can build upon at a later time.
cb02 = True #@param {type:'boolean'}
#@markdown ## What is True about my report?
#@markdown ### I wrote a sufficient number of well-written sentences.
cb03 = True #@param {type:'boolean'}
#@markdown ### My report is free of mechanical infelicities.
cb04 = True #@param {type:'boolean'}
#@markdown ### I used Grammarly (or something better described in my report) to check for MIs.
cb05 = True #@param {type:'boolean'}
#@markdown ### I reported on any connections I found between these problems and something I already know. 
cb06 = True #@param {type:'boolean'}
#@markdown ### I reported who were and what contribution each of my collaborators made.
cb07 = True #@param {type:'boolean'}
#@markdown ## What is True about my explorations?
#@markdown ### I correctly extracted the right modulus.
cb08 = True #@param {type:'boolean'}
#@markdown ### I correctly extracted the right public exponent.
cb09 = True #@param {type:'boolean'}
#@markdown ### I correctly extracted the right private exponent.
cb10 = True #@param {type:'boolean'}
#@markdown ### I correctly extracted the right prime1 value.
cb11 = True #@param {type:'boolean'}
#@markdown ### I correctly extracted the right prime2 value.
cb12 = True #@param {type:'boolean'}
#@markdown ### I correctly extracted the right exponent1 value.
cb13 = True #@param {type:'boolean'}
#@markdown ### I correctly extracted the right exponent2 value.
cb14 = True #@param {type:'boolean'}
#@markdown ### I found the one correct substring of the Title Page of the Book of Mormon that fits all the criteria.
cb15 = True #@param {type:'boolean'}
#@markdown ### I explored the meaning of the exponent1 and exponent2 values.
cb16 = True #@param {type:'boolean'}